### ISSUE TITLE
Tenant id not returned

### ISSUE THREAD

**Issue by DavidPLamas** [flowable/flowable-engine#1511]
Hi,

When using Flowable’s REST API, if you specify either “includeProcessVariables” or “includeTaskLocalVariables” when querying for tasks, the response does not contain the tenant that is associated with that task, it always contains an empty string
If you omit those query params (or set them to false), it works fine: the tenant is correctly set.

This happens for both GET /runtime/tasks and POST /query/tasks.

After some digging, I found that the SQL queries that are fired against the database are different when you want to include variables. It looks like one of them (for example, selectTasksWithRelatedEntitiesByQueryCriteria) is not including the tenant column.

Is this an expected behavior?

David Lamas.

**Comment by martin-grofcik** [comment]
Hi David.

What is your flowable version? (I used master branch to reproduce the issue)
Could you provide simple jUnit test please?

I created simple jUnit test with
````
POST query/tasks
body: {"includeProcessVariables":true,"includeTaskLocalVariables":true}
````

and the response includes tenantId
````
{"data":[{"id":"9","url":"http://localhost:9798/service/runtime/tasks/9","owner":null,"assignee":null,"delegationState":null,"name":"Process task","description":"Process task description","createTime":"2019-01-09T07:48:00.952+01:00","dueDate":null,"priority":50,"suspended":false,"claimTime":null,"taskDefinitionKey":"processTask","scopeDefinitionId":null,"scopeId":null,"scopeType":null,"tenantId":"testTenant","category":null,"formKey":null,"parentTaskId":null,"parentTaskUrl":null,"executionId":"5","executionUrl":"http://localhost:9798/service/runtime/executions/5","processInstanceId":"4","processInstanceUrl":"http://localhost:9798/service/runtime/process-instances/4","processDefinitionId":"oneTaskProcess:1:3","processDefinitionUrl":"http://localhost:9798/service/repository/process-definitions/oneTaskProcess:1:3","variables":[]}],"total":1,"start":0,"sort":"id","order":"asc","size":1}
````

Thanks,
Martin

**Comment by DavidPLamas** [comment]
Hello @martin-grofcik ,

We're currently using Flowable 6.4.0.

Unfortunately, I cannot provide a JUnit test because we have Flowable wrapped inside a OSGi Environment and we haven't figured out a way to write unit tests yet. 

Anyway, we've sort of figured out why this is happening we just don't know to fix it. I'll explain bellow:

```java
public List<Task> executeList(CommandContext commandContext) {
        this.ensureVariablesInitialized();
        this.checkQueryOk();
        List<Task> tasks = null;
        TaskServiceConfiguration taskServiceConfiguration = CommandContextUtil.getTaskServiceConfiguration(commandContext);
        if (taskServiceConfiguration.getTaskQueryInterceptor() != null) {
            taskServiceConfiguration.getTaskQueryInterceptor().beforeTaskQueryExecute(this);
        }

        if (!this.includeTaskLocalVariables && !this.includeProcessVariables && !this.includeIdentityLinks) {
            tasks = CommandContextUtil.getTaskEntityManager(commandContext).findTasksByQueryCriteria(this);
        } else {
            tasks = CommandContextUtil.getTaskEntityManager(commandContext).findTasksWithRelatedEntitiesByQueryCriteria(this);
        }

        if (tasks != null && taskServiceConfiguration.getInternalTaskLocalizationManager() != null && taskServiceConfiguration.isEnableLocalization()) {
            Iterator var4 = tasks.iterator();

            while(var4.hasNext()) {
                Task task = (Task)var4.next();
                taskServiceConfiguration.getInternalTaskLocalizationManager().localize(task, this.locale, this.withLocalizationFallback);
            }
        }

        if (taskServiceConfiguration.getTaskQueryInterceptor() != null) {
            taskServiceConfiguration.getTaskQueryInterceptor().afterTaskQueryExecute(this, tasks);
        }

        return tasks;
    }
```

The second if statement  makes it so two different queries are fired against the database. Digging a bit more:

1. This is the SQL query that is executed when you include task variables or process variables
```SQL
SELECT SUB.* FROM ( 
      select distinct TEMPRES_ID_ as ID_, TEMPRES_REV_ as REV_, TEMPRES_NAME_ as NAME_,
      TEMPRES_PARENT_TASK_ID_ as PARENT_TASK_ID_, TEMPRES_DESCRIPTION_ as DESCRIPTION_,
      TEMPRES_PRIORITY_ as PRIORITY_, TEMPRES_CREATE_TIME_ as CREATE_TIME_,
      TEMPRES_OWNER_ as OWNER_, TEMPRES_ASSIGNEE_ as ASSIGNEE_, TEMPRES_DELEGATION_ as DELEGATION_,
      TEMPRES_FORM_KEY_ as FORM_KEY_, TEMPRES_CLAIM_TIME_ as CLAIM_TIME_, TEMPRES_EXECUTION_ID_ as EXECUTION_ID_,
      TEMPRES_PROC_INST_ID_ as PROC_INST_ID_, TEMPRES_PROC_DEF_ID_ as PROC_DEF_ID_,
      TEMPRES_TASK_DEF_ID_ as TASK_DEF_ID_, TEMPRES_TASK_DEF_KEY_ as TASK_DEF_KEY_,
      TEMPRES_DUE_DATE_ as DUE_DATE_, TEMPRES_SUSPENSION_STATE_ as SUSPENSION_STATE_, TEMPRES_CATEGORY_ as CATEGORY_,
      TEMPRES_IS_COUNT_ENABLED_ as IS_COUNT_ENABLED_,	TEMPRES_VAR_COUNT_ as VAR_COUNT_, TEMPRES_ID_LINK_COUNT_ as ID_LINK_COUNT_, TEMPRES_ID_SUB_TASK_COUNT_ as SUB_TASK_COUNT_   ,
        TEMPVAR_ID_ as VAR_ID_, TEMPVAR_NAME_ as VAR_NAME_, TEMPVAR_TYPE_ as VAR_TYPE_, TEMPVAR_REV_ as VAR_REV_,
        TEMPVAR_PROC_INST_ID_ as VAR_PROC_INST_ID_, TEMPVAR_EXECUTION_ID_ as VAR_EXECUTION_ID_, TEMPVAR_TASK_ID_ as VAR_TASK_ID_,
        TEMPVAR_BYTEARRAY_ID_ as VAR_BYTEARRAY_ID_, TEMPVAR_DOUBLE_ as VAR_DOUBLE_,
        TEMPVAR_TEXT_ as VAR_TEXT_, TEMPVAR_TEXT2_ as VAR_TEXT2_, TEMPVAR_LONG_ as VAR_LONG_ 
       , row_number() over (order by TEMPRES_ID_ asc) rnk FROM ( select distinct  
      RES.ID_ as TEMPRES_ID_, RES.REV_ as TEMPRES_REV_, RES.NAME_ as TEMPRES_NAME_,
      RES.PARENT_TASK_ID_ as TEMPRES_PARENT_TASK_ID_, RES.DESCRIPTION_ as TEMPRES_DESCRIPTION_,
      RES.PRIORITY_ as TEMPRES_PRIORITY_, RES.CREATE_TIME_ as TEMPRES_CREATE_TIME_,
      RES.OWNER_ as TEMPRES_OWNER_, RES.ASSIGNEE_ as TEMPRES_ASSIGNEE_, RES.DELEGATION_ as TEMPRES_DELEGATION_,
      RES.FORM_KEY_ as TEMPRES_FORM_KEY_, RES.CLAIM_TIME_ as TEMPRES_CLAIM_TIME_, RES.EXECUTION_ID_ as TEMPRES_EXECUTION_ID_,
      RES.PROC_INST_ID_ as TEMPRES_PROC_INST_ID_, RES.PROC_DEF_ID_ as TEMPRES_PROC_DEF_ID_,
      RES.TASK_DEF_ID_ as TEMPRES_TASK_DEF_ID_, RES.TASK_DEF_KEY_ as TEMPRES_TASK_DEF_KEY_,
      RES.DUE_DATE_ as TEMPRES_DUE_DATE_, RES.SUSPENSION_STATE_ as TEMPRES_SUSPENSION_STATE_, RES.CATEGORY_ as TEMPRES_CATEGORY_,
      RES.IS_COUNT_ENABLED_ as TEMPRES_IS_COUNT_ENABLED_, RES.VAR_COUNT_ as TEMPRES_VAR_COUNT_, RES.ID_LINK_COUNT_ as TEMPRES_ID_LINK_COUNT_, RES.SUB_TASK_COUNT_ as TEMPRES_ID_SUB_TASK_COUNT_       ,
        VAR.ID_ as TEMPVAR_ID_, VAR.NAME_ as TEMPVAR_NAME_, VAR.TYPE_ as TEMPVAR_TYPE_, VAR.REV_ as TEMPVAR_REV_,
        VAR.PROC_INST_ID_ as TEMPVAR_PROC_INST_ID_, VAR.EXECUTION_ID_ as TEMPVAR_EXECUTION_ID_, VAR.TASK_ID_ as TEMPVAR_TASK_ID_,
        VAR.BYTEARRAY_ID_ as TEMPVAR_BYTEARRAY_ID_, VAR.DOUBLE_ as TEMPVAR_DOUBLE_,
        VAR.TEXT_ as TEMPVAR_TEXT_, VAR.TEXT2_ as TEMPVAR_TEXT2_, VAR.LONG_ as TEMPVAR_LONG_  
    from ACT_RU_TASK RES
          left outer join ACT_RU_VARIABLE VAR ON RES.PROC_INST_ID_ = VAR.EXECUTION_ID_ and VAR.TASK_ID_ is null
     )RES ) SUB WHERE SUB.rnk >= ? AND SUB.rnk < ?
```
2. And this is the SQL query that is executed when you dont include variables
```SQL
SELECT SUB.* FROM ( 
    select distinct RES.*    , row_number() over (order by RES.ID_ asc) rnk FROM ( select distinct RES.*  
        from ACT_RU_TASK RES )RES ) SUB WHERE SUB.rnk >= ? AND SUB.rnk < ?
```

It looks like the first is not including the tenant. This might be a issue that is already resolved in the master branch. 
Also, I should note I manually tested the use case you implemented using JUnit and it does not work as you would expect.

**Comment by martin-grofcik** [comment]
Hi David.

Thanks for the analysis. I was able to reproduce the issue. The problem was exactly in this query and it was missing tenantId. My previous test used h2 DB (as a default DB), and you are using mssql or db2, right? I already created a pull request to fix this issue.
https://github.com/flowable/flowable-engine/pull/1522

Regards
Martin

**Comment by martin-grofcik** [comment]
Hi David,

It is hard to predict the release, but you can build libraries from the sources and cherry-pick this fix.

Regards
Martin

**Comment by wberges** [comment]
Hello,
Certainly a stupid question, but why don't you publish regularly a SNAPSHOT of the current version? It would allow the users to test also on their side the current candidate and certainly to detect bugs early? It would be easier than forcing each user to rebuild it, no? 
But there's certainly a good reason to not publish it. 😉
Best regards

**Comment by DavidPLamas** [comment]
Is there any way that this issue can be fixed and published in the latest release (6.4.0) ? 
The reason I'm asking this is because I work with a big team and we have an internal maven repository (on NEXUS) where we store both internal and external projects. We can't risk using the SNAPSHOT version because we have software that is ready to enter production phase.

If the answer is no, do you think I cloning the branch 6.4.0 and making the same changes in pull request #1522 (locally) would fix the issue?

Thank you,
David Lamas.

**Comment by DavidPLamas** [comment]
Oh, that's some great news. 
One (or even two) weeks is something that we are willing to wait. 

Thank you for the quick response,
David Lamas.

**Comment by PascalSchumacher** [comment]
I guess this can be closed as the fix was merged and flowable `6.4.1` was released?
